### PR TITLE
SONARJS-672 Tune rule StringConcatenationCheck (S3512)

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

**Comment by pynicolas** [comment]
@vilchik-elena It seems that we have false negatives because `visitBinaryExpression` does not call its `super` implementation.

**Comment by vilchik-elena** [comment]
@pynicolas No, we shouldn't. We visit operands in `checkBinaryExpression`

**Comment by pynicolas** [comment]
@vilchik-elena You're right. Sorry for the noise.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java [hunk 1]
```diff
@@ -62,28 +62,50 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
...
-        }
-      }
+    int operandsNum = operandList.size();
+    int literalsNum = stringLiteralsNumber(operandList);
+
+    if (operandsNum > 2 && literalsNum > 0) {
+      int shortLiteralsNum = smallLiteralsNumber(operandList);
 
-      return literalsNum > 0 && shortLiteralsNum <= literalsNum / 2.;
+      boolean isOneLiteralTreeOperandsCase = operandsNum == 3 && literalsNum == 1;
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java]: @vilchik-elena Typo? Shouldn't we have a "h" as in "Three"?

Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java]: @pynicolas Thanks!:)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java [hunk 2]
```diff
@@ -62,28 +62,50 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
...
+      boolean atLeastHalfLongLiterals = shortLiteralsNum <= literalsNum / 2.;
+      boolean notTooManyNonLiterals = literalsNum - (operandsNum - literalsNum) >= -1;
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java]: @vilchik-elena Is it equivalent to `literalsNum > nonLiteralsNum`? That would be clearer I think.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java (hunk 1 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java]: @pynicolas fixed 53df9786c8f7d4ec1c31ecf5923a0c9b60933ed1

File: javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java (hunk 2 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java]: @pynicolas No, it's not equivalent. It's the same is `literalsNum > nonLiteralsNum - 2`. Is it clearer?:)

Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/StringConcatenationCheck.java]: @vilchik-elena A bit.

**Pull request merged**